File: web/lib/api/pms-sandbox.ts
import { request } from "@/lib/api/client"

export type SandboxDataset = { id: number; storeId: number; version: number; createdAt: string }
export type SandboxRoomType = { id: number; name: string; rank: number; priceCents: number; enabled: boolean }
export type SandboxRoom = {
  id: number; roomTypeId: number; number: string; floor: string
  cleanStatus: string; description: string; enabled: boolean
}
export type SandboxOrder = {
  id: number; number: string; guestName: string; phone: string
  roomTypeId: number; roomTypeName: string; roomId: number; roomNumber: string
  checkIn: string; checkOut: string; payableCents: number
  includesBreakfast: boolean; status: string; version: number
}
export type SandboxGrade = {
  id: number; name: string; benefits: string[]; birthdayBenefits: string[]
  freeUpgradeMaxRank: number; enabled: boolean
}
export type SandboxMember = {
  id: number; name: string; phone: string; gradeId: number; gradeName: string
  birthday: string; validUntil: string; enabled: boolean
}
export type SandboxRule = {
  id: number; code: string; name: string; text: string; action: string
  amountCents: number; checkoutTime: string; minimumGradeId: number; enabled: boolean
  validFrom?: string | null; validUntil?: string | null
}
export type SandboxResource = {
  id: number; code: string; name: string; token: string; sourceMessageId: number
  messageType: string; enabled: boolean; cardPayload?: string
}
export type SandboxBinding = { id: number; customerId: number; orderId: number; memberId: number }
export type SandboxOperation = {
  id: number; provider: string; storeId: number; datasetId: number
  conversationId: number; sourceMessageId: number; previewMessageId: number
  confirmationMessageId: number; operationType: string; status: string
  previewText: string; resultText: string; errorMessage: string
  plan?: { before: SandboxOrder; after: SandboxOrder; addedCents: number; commitment: string }
  expiresAt?: string; createdAt: string; deliveryStatus: string
}
export type SandboxOption = { value: string; label: string }
export type SandboxCustomerOption = { id: number; name: string; conversationId: number }
export type SandboxWorkspace = {
  label: string; active: boolean; dataset?: SandboxDataset
  roomTypes: SandboxRoomType[]; rooms: SandboxRoom[]; orders: SandboxOrder[]
  grades: SandboxGrade[]; members: SandboxMember[]; rules: SandboxRule[]
  resources: SandboxResource[]; bindings: SandboxBinding[]; operations: SandboxOperation[]
  options: Record<string, SandboxOption[]>
}
export type SandboxSave = {
  roomType?: SandboxRoomType; room?: SandboxRoom; order?: SandboxOrder
  grade?: SandboxGrade; member?: SandboxMember; rule?: SandboxRule; resource?: SandboxResource
}

const base = "/api/dashboard/pms-sandbox"
function post<T>(path: string, body: unknown) {
  return request<T>(`${base}/${path}`, { method: "POST", body: JSON.stringify(body) })
}
export function fetchSandboxStores() {
  return request<{ id: number; name: string }[]>(`${base}/stores`)
}
export function fetchSandboxWorkspace(storeId: number) {
  return request<SandboxWorkspace>(`${base}/list?storeId=${storeId}`)
}
export function fetchSandboxCustomers(storeId: number) {
  return request<SandboxCustomerOption[]>(`${base}/customers?storeId=${storeId}`)
}
export function initializeSandbox(storeId: number) {
  return post<SandboxWorkspace>("initialize", { storeId })
}
export function resetSandbox(storeId: number, dataset: SandboxDataset) {
  return post<SandboxWorkspace>("reset", { storeId, datasetId: dataset.id, version: dataset.version })
}
export function saveSandbox(storeId: number, dataset: SandboxDataset, value: SandboxSave) {
  return post<SandboxWorkspace>("update", { storeId, datasetId: dataset.id, version: dataset.version, ...value })
}
export function bindSandboxCustomer(storeId: number, dataset: SandboxDataset, value: Omit<SandboxBinding, "id">) {
  return post<SandboxWorkspace>("bind", { storeId, datasetId: dataset.id, version: dataset.version, ...value })
}
export function cancelSandboxOperation(storeId: number, operationId: number) {
  return post<void>("cancel", { storeId, operationId })
}

File: internal/models/pms_sandbox.go
package models

import "time"

type PMSSandboxStore struct {
	ID              int64     `gorm:"primaryKey;autoIncrement"`
	StoreID         int64     `gorm:"type:bigint;not null;uniqueIndex"`
	ActiveDatasetID int64     `gorm:"type:bigint;not null;default:0"`
	LockVersion     int64     `gorm:"type:bigint;not null;default:0"`
	UpdatedAt       time.Time `gorm:"type:datetime;not null"`
}

type PMSSandboxDataset struct {
	ID        int64     `gorm:"primaryKey;autoIncrement"`
	StoreID   int64     `gorm:"type:bigint;not null;index"`
	Version   int64     `gorm:"type:bigint;not null;default:1"`
	CreatedAt time.Time `gorm:"type:datetime;not null"`
}

type PMSSandboxRoomType struct {
	ID         int64  `gorm:"primaryKey;autoIncrement"`
	StoreID    int64  `gorm:"type:bigint;not null;index"`
	DatasetID  int64  `gorm:"type:bigint;not null;index"`
	Name       string `gorm:"type:varchar(80);not null"`
	Rank       int    `gorm:"type:int;not null"`
	PriceCents int64  `gorm:"type:bigint;not null"`
	Enabled    bool   `gorm:"not null"`
}

type PMSSandboxRoom struct {
	ID          int64  `gorm:"primaryKey;autoIncrement"`
	StoreID     int64  `gorm:"type:bigint;not null;index"`
	DatasetID   int64  `gorm:"type:bigint;not null;uniqueIndex:uk_sandbox_room_number"`
	RoomTypeID  int64  `gorm:"type:bigint;not null;index"`
	Number      string `gorm:"type:varchar(40);not null;uniqueIndex:uk_sandbox_room_number"`
	Floor       string `gorm:"type:varchar(40);not null"`
	CleanStatus string `gorm:"type:varchar(20);not null"`
	Description string `gorm:"type:text"`
	Enabled     bool   `gorm:"not null"`
}

type PMSSandboxOrder struct {
	ID                int64     `gorm:"primaryKey;autoIncrement"`
	StoreID           int64     `gorm:"type:bigint;not null;index"`
	DatasetID         int64     `gorm:"type:bigint;not null;uniqueIndex:uk_sandbox_order_number"`
	Number            string    `gorm:"type:varchar(80);not null;uniqueIndex:uk_sandbox_order_number"`
	GuestName         string    `gorm:"type:varchar(80);not null"`
	Phone             string    `gorm:"type:varchar(30);not null;index"`
	RoomTypeID        int64     `gorm:"type:bigint;not null;index"`
	RoomID            int64     `gorm:"type:bigint;not null;index"`
	CheckIn           time.Time `gorm:"type:datetime;not null;index"`
	CheckOut          time.Time `gorm:"type:datetime;not null;index"`
	PayableCents      int64     `gorm:"type:bigint;not null"`
	IncludesBreakfast bool      `gorm:"not null"`
	Status            string    `gorm:"type:varchar(30);not null;index"`
	Version           int64     `gorm:"type:bigint;not null;default:1"`
	UpdatedAt         time.Time `gorm:"type:datetime;not null"`
}

type PMSSandboxGrade struct {
	ID                 int64  `gorm:"primaryKey;autoIncrement"`
	StoreID            int64  `gorm:"type:bigint;not null;index"`
	DatasetID          int64  `gorm:"type:bigint;not null;index"`
	Name               string `gorm:"type:varchar(80);not null"`
	BenefitsJSON       string `gorm:"type:text"`
	BirthdayJSON       string `gorm:"type:text"`
	FreeUpgradeMaxRank int    `gorm:"type:int;not null"`
	Enabled            bool   `gorm:"not null"`
}

type PMSSandboxMember struct {
	ID         int64     `gorm:"primaryKey;autoIncrement"`
	StoreID    int64     `gorm:"type:bigint;not null;index"`
	DatasetID  int64     `gorm:"type:bigint;not null;index"`
	Name       string    `gorm:"type:varchar(80);not null"`
	Phone      string    `gorm:"type:varchar(30);not null;index"`
	GradeID    int64     `gorm:"type:bigint;not null"`
	Birthday   string    `gorm:"type:varchar(10);not null"`
	ValidUntil time.Time `gorm:"type:datetime;not null"`
	Enabled    bool      `gorm:"not null"`
}

type PMSSandboxRule struct {
	ID             int64      `gorm:"primaryKey;autoIncrement"`
	StoreID        int64      `gorm:"type:bigint;not null;index"`
	DatasetID      int64      `gorm:"type:bigint;not null;uniqueIndex:uk_sandbox_rule_code"`
	Code           string     `gorm:"type:varchar(80);not null;uniqueIndex:uk_sandbox_rule_code"`
	Name           string     `gorm:"type:varchar(100);not null"`
	Text           string     `gorm:"type:text"`
	Action         string     `gorm:"type:varchar(40);not null"`
	AmountCents    int64      `gorm:"type:bigint;not null"`
	CheckoutTime   string     `gorm:"type:varchar(10);not null"`
	MinimumGradeID int64      `gorm:"type:bigint;not null;default:0"`
	Enabled        bool       `gorm:"not null"`
	ValidFrom      *time.Time `gorm:"type:datetime"`
	ValidUntil     *time.Time `gorm:"type:datetime"`
}

type PMSSandboxResource struct {
	ID              int64  `gorm:"primaryKey;autoIncrement"`
	StoreID         int64  `gorm:"type:bigint;not null;index"`
	DatasetID       int64  `gorm:"type:bigint;not null;uniqueIndex:uk_sandbox_resource_code"`
	Code            string `gorm:"type:varchar(80);not null;uniqueIndex:uk_sandbox_resource_code"`
	Name            string `gorm:"type:varchar(100);not null"`
	Token           string `gorm:"type:text"`
	CardPayload     string `gorm:"type:text"`
	MessageType     string `gorm:"type:varchar(30);not null;default:''"`
	SourceMessageID int64  `gorm:"type:bigint;not null;default:0"`
	Enabled         bool   `gorm:"not null"`
}

type PMSSandboxBinding struct {
	ID         int64 `gorm:"primaryKey;autoIncrement"`
	StoreID    int64 `gorm:"type:bigint;not null;index"`
	DatasetID  int64 `gorm:"type:bigint;not null;uniqueIndex:uk_sandbox_customer_binding"`
	CustomerID int64 `gorm:"type:bigint;not null;uniqueIndex:uk_sandbox_customer_binding"`
	OrderID    int64 `gorm:"type:bigint;not null"`
	MemberID   int64 `gorm:"type:bigint;not null"`
}
